extensions/roc-plugin-repo/src/commands/bootstrap.js

```diff
@@ -1,135 +1,181 @@
-    .map(dependency => `${binary} link ${dependency}`)
-    .join(' && ')}`;
-};
-
-const removeDependencies = (dependencies = {}, localDependencies) => {
+import onExit from 'signal-exit';
+import readPkg from 'read-pkg';
+import writePkg from 'write-pkg';
+import Listr from 'listr';
+import { createLink, createBinaryLink } from './utils/install';
+import generateStatus from '../semver/generateStatus';
+import { getNextVersions, createVersionsDoesNotMatch } from '../semver/utils';
+
+const removeDependencies = (
+  dependencies = {},
+  localDependencies,
+  ignoreSemVer,
+) => {
   const newDependencies = {};
 
+  const removeDependenciesThatShouldBeLinked = createVersionsDoesNotMatch(
+    localDependencies,
+    dependencies,
+    ignoreSemVer,
+  );
+
   Object.keys(dependencies)
-    .filter(dependency => !localDependencies.includes(dependency))
+    .filter(removeDependenciesThatShouldBeLinked)
     .forEach(dependency => {
       newDependencies[dependency] = dependencies[dependency];
     });
 
   return newDependencies;
 };
 
-const install = (project, extra, binary, localDependencies, useInstall) => {
+const install = async (
+  project,
+  binary,
+  localDependencies,
+  { ignoreSemVer, verbose },
+) => {
   const pathToPackageJSON = path.join(project.path, 'package.json');
-  const packageJSON = require(pathToPackageJSON); // eslint-disable-line
-  const newPackageJSON = Object.assign({}, packageJSON);
+  const pathToPackageJSONBackup = `${pathToPackageJSON}.backup`;
+  const packageJSON = readPkg.sync(pathToPackageJSON);
+  const tempPackageJSON = Object.assign({}, packageJSON);
 
-  newPackageJSON.dependencies = removeDependencies(
+  tempPackageJSON.dependencies = removeDependencies(
     packageJSON.dependencies,
-    Object.keys(localDependencies),
+    localDependencies,
+    ignoreSemVer,
   );
-  newPackageJSON.devDependencies = removeDependencies(
+  tempPackageJSON.devDependencies = removeDependencies(
     packageJSON.devDependencies,
-    Object.keys(localDependencies),
+    localDependencies,
+    ignoreSemVer,
   );
 
-  fs.writeFileSync(
-    pathToPackageJSON,
-    `${JSON.stringify(newPackageJSON, null, 2)}\n`,
-  );
+  await fs.rename(pathToPackageJSON, pathToPackageJSONBackup);
 
-  log.info(`Installing dependencies and linking ${project.name}`);
+  await writePkg(pathToPackageJSON, tempPackageJSON);
 
   const restorePackageJSON = () =>
-    fs.writeFileSync(
-      pathToPackageJSON,
-      `${JSON.stringify(packageJSON, null, 2)}\n`,
-    );
-
-  return execute(
-    `cd ${project.path}${linkExtra(
-      extra,
-      binary,
-    )} && ${binary} install${useInstall ? '' : ` && ${binary} link`}`,
-  ).then(restorePackageJSON, error => {
-    restorePackageJSON();
-    throw error;
-  });
+    fs.renameSync(pathToPackageJSONBackup, pathToPackageJSON);
+
+  const unregister = onExit(restorePackageJSON);
+
+  return execute(`cd ${project.path} && ${binary} install`, {
+    silent: !verbose,
+  }).then(
```

Comment:
I will rewrite this, thanks for pointing it out. Most of the `then`s are a rest before `async await` was used.